python/semantic_kernel/connectors/ai/open_ai/services/open_ai_chat_completion_base.py

```diff
@@ -73,39 +87,68 @@ async def get_chat_message_contents(
+            if kernel is None or arguments is None:
+                raise ServiceInvalidExecutionSettingsError(
+                    "The kernel and kernel arguments are required for auto invoking OpenAI tool calls."
+                )
+            if settings.number_of_responses > 1:
+                raise ServiceInvalidExecutionSettingsError(
+                    "Auto-invocation of tool calls may only be used with a "
+                    "OpenAIChatPromptExecutions.number_of_responses of 1."
+                )
+
         # behavior for non-function calling or for enable, but not auto-invoke.
-        settings = self._prepare_settings(settings, chat_history, stream_request=False, kernel=kernel)
+        self._prepare_settings(settings, chat_history, stream_request=False, kernel=kernel)
         if settings.function_call_behavior is None or (
             settings.function_call_behavior and not settings.function_call_behavior.auto_invoke_kernel_functions
         ):
             return await self._send_chat_request(settings)
 
         # loop for auto-invoke function calls
-        for _ in range(settings.function_call_behavior.max_auto_invoke_attempts):
+        for request_index in range(settings.function_call_behavior.max_auto_invoke_attempts):
             completions = await self._send_chat_request(settings)
-            if all(
-                not isinstance(item, FunctionCallContent) for completion in completions for item in completion.items
-            ):
+            # there is only one chat message, this was checked earlier
+            chat_history.add_message(message=completions[0])
+            # get the function call contents from the chat message
+            function_calls = [item for item in chat_history.messages[-1].items if isinstance(item, FunctionCallContent)]
+            if (fc_count := len(function_calls)) == 0:
+                return completions
+
+            logger.info(f"processing {fc_count} tool calls in parallel.")
+            try:
+                # this function either updates the chat history with the function call results
+                # or raises a the InvokeTermination through a filter,
+                # in which case the loop will break and the function calls are returned.
+                await asyncio.gather(
+                    *[
+                        self._process_function_call(
+                            function_call=function_call,
+                            chat_history=chat_history,
+                            kernel=kernel,
+                            arguments=arguments,
+                            function_call_count=fc_count,
+                            request_index=request_index,
+                            function_call_behavior=settings.function_call_behavior,
+                        )
+                        for function_call in function_calls
+                    ],
+                )
+            except InvokeTermination:
```

Comment:
It may be good to show this as a concept example? I know a lot of people are wanting this new functionality.

Reply:
The auto_function_invocation_filter example uses this!
Set context.terminate to True and then the InvokeTermination is thrown, that is used just because we are going between two functions, users do not see that one